commit e80e672819cc81807628132501c1bda426c54d5f
Author: Keiichi Fujino <kfujino@apache.org>
Date:   Tue Jun 21 06:21:40 2016 +0000

    If the ping message has been received at the AbstractReplicatedMap#leftOver method, ensure that notify the member is alive than ignore it.
    
    git-svn-id: https://svn.apache.org/repos/asf/tomcat/trunk@1749445 13f79535-47bb-0310-9956-ffa450edef68

diff --git a/java/org/apache/catalina/tribes/tipis/AbstractReplicatedMap.java b/java/org/apache/catalina/tribes/tipis/AbstractReplicatedMap.java
index 339d075..79320c5 100644
--- a/java/org/apache/catalina/tribes/tipis/AbstractReplicatedMap.java
+++ b/java/org/apache/catalina/tribes/tipis/AbstractReplicatedMap.java
@@ -616,7 +616,8 @@ public abstract class AbstractReplicatedMap<K,V>
             mapmsg.deserialize(getExternalLoaders());
             if (mapmsg.getMsgType() == MapMessage.MSG_START) {
                 mapMemberAdded(mapmsg.getPrimary());
-            } else if (mapmsg.getMsgType() == MapMessage.MSG_INIT) {
+            } else if (mapmsg.getMsgType() == MapMessage.MSG_INIT
+                    || mapmsg.getMsgType() == MapMessage.MSG_PING) {
                 memberAlive(mapmsg.getPrimary());
             } else {
                 // other messages are ignored.
diff --git a/webapps/docs/changelog.xml b/webapps/docs/changelog.xml
index d91f9c1..dea47c5 100644
--- a/webapps/docs/changelog.xml
+++ b/webapps/docs/changelog.xml
@@ -102,6 +102,11 @@
       <add>
         Add log message when the ping has timed-out. (kfujino)
       </add>
+      <fix>
+        If the ping message has been received at the
+        <code>AbstractReplicatedMap#leftOver</code> method, ensure that notify
+        the member is alive than ignore it. (kfujino)
+      </fix>
     </changelog>
   </subsection>
   <subsection name="Other">
